### PR TITLE
Fix L1 fat event filters 81X

### PR DESCRIPTION
**Pull request by jimbrooke**
81X version of #16765

Fixes for two filters used to select L1T "fat events", to use of the 64-bit event number from TCDS FED word 11 (https://twiki.cern.ch/twiki/bin/viewauth/CMS/TcdsEventRecord#TCDS_Data_Block_Definition_curre) instead of the 24-bit event number from common DAQ header.

In the case of HLTL1NumberFilter, use of the common DAQ header (ie. existing functionality) is maintained as default.

### CODE REVIEW

**Comment by rekovic** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by rekovic** [comment]
+1

File: HLTrigger/special/interface/HLTL1NumberFilter.h [hunk 1]
```diff
@@ -55,6 +55,8 @@ class HLTL1NumberFilter : public edm::global::EDFilter<> {
   const int fedId_;
   /// if invert_=true, invert that event accept decision
   const bool invert_;
+  /// if useTCDS=true, use 64-bit Event Number from TCDS record (FED 1024) word 11
+  bool useTCDS_;
```
Comment (fwyzard) [inline HLTrigger/special/interface/HLTL1NumberFilter.h]: can you make this `const` ?

File: HLTrigger/special/src/HLTL1NumberFilter.cc [hunk 2]
```diff
@@ -37,8 +39,13 @@ HLTL1NumberFilter::HLTL1NumberFilter(const edm::ParameterSet& config) :
...
   period_( config.getParameter<unsigned int>("period") ),
   fedId_(  config.getParameter<int>("fedId") ),
-  invert_( config.getParameter<bool>("invert") )
+  invert_( config.getParameter<bool>("invert") ),
+  useTCDS_( config.getParameter<bool>("useTCDSEventNumber") )
 {
+
+  // only try and use TCDS event number if the FED ID 1024 is selected
+  if (fedId_!=1024) useTCDS_ = false;
+
```
Comment (fwyzard) [inline HLTrigger/special/src/HLTL1NumberFilter.cc]: can you move this into the office initialiser above ?

File: HLTrigger/special/src/HLTL1NumberFilter.cc [hunk 3]
```diff
@@ -76,6 +84,10 @@ HLTL1NumberFilter::filter(edm::StreamID, edm::Event& iEvent, const edm::EventSet
     if (data.data() && data.size() > 0){
       FEDHeader header(data.data()) ;
       if (period_!=0) accept = ( ( (header.lvl1ID())%period_ ) == 0 );
+      if (useTCDS_ && period_!=0) {
+	evf::evtn::TCDSRecord record((unsigned char *)data.data());
+	accept = ( (record.getHeader().getData().header.triggerCount % period_) == 0 );
+      }
```
Comment (fwyzard) [inline HLTrigger/special/src/HLTL1NumberFilter.cc]: can you reword this so that there is a single check on the counter ?

**Review by fwyzard — CHANGES_REQUESTED** [review]

**Comment by fwyzard** [comment]
are my retreats for changes even visible ?

**Comment by davidlange6** [comment]
hi @jimbrooke @rekovic - please open/approve a 90x version of this pr. Thanks

**Comment by fwyzard** [comment]
hold

please address the comments above

**Comment by jimbrooke** [comment]
I think it would be more efficient if someone who is actually familiar with HLT coding rules fix this bug.  I certainly don't have time for this.  Please see https://github.com/cms-l1t-offline/cmssw/pull/438 for an example of the 3 line change that is required.

**Comment by fwyzard** [comment]
The HLT coding rules are "do not write crappy code".
What makes you think that the HLT people have the time to fix the L1 code ?

**Comment by fwyzard** [comment]
please see #16782

**Comment by fwyzard** [comment]
-1

**Comment by rekovic** [comment]
superseded by #16782

**Pull request closed without merging**